File: programmers/algorithm_category/unanchoi/bf/피로도.py
from itertools import permutations

def solution(k, dungeons):
    answer_list = []

    all = []
    for p in permutations(dungeons, len(dungeons)):
        all.append(list(p))
    print(all)

    i = 0
    while i < len(all):
        answer = 0
        피로도 = k
        #print(dungeon)
        #print(len(dungeon))
        for d in all[i]:
            최소필요피로도 = d[0]
            소모피로도 = d[1]

            if 피로도 < 최소필요피로도:
                break
            else:
                피로도 -= 소모피로도
                answer += 1
           # print(피로도)
        answer_list.append(answer)
        i += 1
    print(answer_list)
    return max(answer_list)

print(solution(80, [[80,20],[50,40],[30,10]]))

File: programmers/algorithm/level2/sung-silver/피로도.java
class Solution {
    static boolean[] visited;
    static int answer;
    public int solution(int k, int[][] dungeons) {
        // dungeons를 탐색하여 가능한 갯수를 각각 구한다
        visited = new boolean[dungeons.length];
        dfs(0,k,dungeons);
        return answer;
    }

    private void dfs(int depth, int health, int[][] dungeons){
        for (int i = 0; i < dungeons.length; i++){
            if (visited[i] || dungeons[i][0] > health) {
                continue;
            }
            visited[i] = true;
            dfs(depth + 1, health - dungeons[i][1], dungeons);
            visited[i] = false;
        }
        answer = Math.max(answer, depth);
    }
}
